qemu/tests/usb_smartcard_sharing.py

```diff
@@ -0,0 +1,210 @@
+        logging.info(out)
+        if s:
+            status = False
+            err_msg = 'Fail to get installed group list in guest, '
+            err_msg += 'output: %s' % out
+            return (status, err_msg)
+        for group in required_groups:
+            if group not in out:
+                groupinstall_cmd = "yum groupinstall -y '%s'" % group
+                s, o = session.cmd_status_output(groupinstall_cmd,
+                                                 timeout=timeout)
+                if s:
+                    status = False
+                    err_msg = "Fail to install group '%s' in guest, " % group
+                    err_msg += "output: %s" % o
+                    return (status, err_msg)
+        o = session.cmd_output('systemctl status pcscd')
+        logging.info(o)
+        if 'running' not in o:
+            s, o = session.cmd_status_output('sytemctl restart pcscd')
+            if s:
+                status = False
+                err_msg = 'Fail to start pcscd in guest, '
+                err_msg += 'output: %s' % out
+                return (status, err_msg)
+        return (status, err_msg)
+
+    def _check_sc_in_guest():
+        status = True
+        err_msg = ''
+        o = session.cmd_output('lsusb')
+        logging.info(o)
+        if ccid_info not in o:
+            status = False
+            err_msg = 'USB CCID device is not present in guest.'
+            return (status, err_msg)
+        list_certs_cmd = 'pkcs11-tool --list-objects --type cert'
+        s, o = session.cmd_status_output(list_certs_cmd)
+        logging.info(o)
+        if s:
+            status = False
+            err_msg = 'Fail to list certificates on the smartcard.'
+            return (status, err_msg)
+        return (status, err_msg)
+
+    def _usbsc_sharing_preprocess():
+        def _generate_usbsc_sharing_cmdline():
+            extra_params = ''
+            backend = usbsc_params.get('chardev_backend', 'spicevmc')
+            name = usbsc_params.get('chardev_name', 'smartcard')
+            chardev_id = usbsc_params.get("chardev_id",
+                                          "chardev_%s" % usbscdev_name)
+            chardev_params = Params({'backend': backend, 'id': chardev_id})
+            chardev_params['name'] = name
+            chardev_params['debug'] = usbsc_params.get('chardev_debug')
+            chardev = qdevices.CharDevice(chardev_params, chardev_id)
+            usbsc_dev = qdevices.QDevice('ccid-card-passthru',
+                                         aobject=usbscdev_name)
+            usbsc_dev.set_param('id', 'usb-%s' % usbscdev_name)
+            usbsc_dev.set_param('chardev', chardev_id)
+            extra_params += ' '.join([chardev.cmdline(),
+                                      usbsc_dev.cmdline()])
+            return extra_params
+        extra_params = _generate_usbsc_sharing_cmdline()
+        params["extra_params"] = extra_params
```

Comment:
From L122 ~L140, i think it can be replaced with the function `add_smartcard` in the framework.
Refer update the cfg with following params:
```
        smartcard = 'yes'
        smartcard_chardev = 'spicevmc'
        smartcard_id = 'chardev_${usbscdev_name}'
```
The command line for those configuration is:
``` 
  -chardev spicevmc,id=chardev_sc1,name=smartcard \
  -device ccid-card-passthru,chardev=chardev_sc1 \
```
The only difference is miss id `id=usb-sc1` for device `ccid-card-passthru`.

Reply:
Missed one issue~   `add_smartcard` will add the `usb-ccid` device too, while `usb_devices += " ccid1"` and `usb_type = "usb-ccid"` already define a `usb-ccid` device.
From my testing, remove `usb_devices += " ccid1"` can work well.   Let `add_smartcard` to define all the related devices is enough.

Reply:
Missed one info, `add_smartcard` will add the `usb-ccid` device while `usb_devices += " ccid1"` define a `usb-ccid` device too. 

Remove `usb_devices += " ccid1"` and let `add_smarcard` to add all the related device is enough i think, and test pass in my side.

Reply:
Yes, you're right, thanks!